File: 0_1.baekjoon/brute-force/java/src/BOJ18808.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ18808 {
    private static int N, M, K;
    private static int[][][] stickers;
    private static int[][] laptop;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        laptop = new int[N][M];
        stickers = new int[K][][];

        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int r = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());

            stickers[i] = new int[r][c];

            for (int j = 0; j < r; j++) {
                st = new StringTokenizer(br.readLine());
                for (int k = 0; k < c; k++) {
                    stickers[i][j][k] = Integer.parseInt(st.nextToken());
                }
            }
        }

        System.out.println(countAttachedBlocks());
    }

    private static int countAttachedBlocks() {
        for (int[][] sticker : stickers) {
            for (int i = 0; i < 4; i++) {
                if (attach(sticker)) {
                    break;
                }
                sticker = rotate(sticker);
            }
        }

        return count();
    }

    private static int count() {
        int count = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (laptop[i][j] != 0) {
                    count++;
                }
            }
        }

        return count;
    }

    private static int[][] rotate(int[][] sticker) {
        int m = sticker[0].length;
        int n = sticker.length;
        int[][] temp = new int[m][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                temp[j][n - i - 1] = sticker[i][j];
            }
        }

        return temp;
    }

    private static boolean attach(int[][] sticker) {

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (isAttachable(i, j, sticker)) {
                    attach(i, j, sticker);
                    return true;
                }
            }
        }
        return false;
    }

    private static void attach(int row, int col, int[][] sticker) {
        int n = sticker.length;
        int m = sticker[0].length;

        for (int i = row; i < row + n; i++) {
            for (int j = col; j < col + m; j++) {
                if (sticker[i - row][j - col] == 1) {
                    laptop[i][j] = 1;
                }
            }
        }
    }

    private static boolean isAttachable(int row, int col, int[][] sticker) {

        int n = sticker.length;
        int m = sticker[0].length;

        if (row + n > N || col + m > M) {
            return false;
        }

        for (int i = row; i < row + n; i++) {
            for (int j = col; j < col + m; j++) {
                if (sticker[i - row][j - col] == 1 && laptop[i][j] == 1) {
                    return false;
                }
            }
        }

        return true;
    }

}

File: 0_1.baekjoon/brute-force/kotlin/src/BOJ18808.kt
private var N = 0
private var M = 0
private var K = 0

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m, k) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    K = k

    val laptop = Array(N) { IntArray(M) { 0 } }
    repeat(k) {
        val(r, c) = readLine().split(" ").map { it.toInt() }
        var sticker = Array(r) { readLine().split(" ").map { it.toInt() }.toIntArray() }

        for (i in 0 until 4) {
            if (attachTo(sticker, laptop)) {
                break
            }
            sticker = rotate(sticker)

        }
    }

    print(countAttachedBlocks(laptop))
}

private fun rotate(sticker: Array<IntArray>): Array<IntArray> {
    val n = sticker.size
    val m = sticker[0].size

    val temp = Array(m) {IntArray(n) { 0 } }

    for (i in 0 until n) {
        for (j in 0 until m) {
            temp[j][n - i - 1] = sticker[i][j]
        }
    }

    return temp
}

private fun attachTo(sticker: Array<IntArray>, laptop: Array<IntArray>): Boolean {
    for (i in 0 until N) {
        for (j in 0 until M) {
            if (isAttachableTo(i, j, sticker, laptop)) {
                attachTo(i, j, sticker, laptop)
                return true
            }
        }
    }
    return false
}

private fun isAttachableTo(row: Int, col: Int, sticker: Array<IntArray>, laptop: Array<IntArray>): Boolean {

    val n = sticker.size
    val m = sticker[0].size

    if (row + n > N || col + m > M) {
        return false
    }

    for (i in row until row + n) {
        for (j in col until col + m) {
            if (sticker[i - row][j - col] == 1 && laptop[i][j] == 1) {
                return false
            }
        }
    }

    return true
}

private fun attachTo(row: Int, col: Int, sticker: Array<IntArray>, laptop: Array<IntArray>) {
    val n = sticker.size
    val m = sticker[0].size

    for (i in row until row + n) {
        for (j in col until col + m) {
            if (sticker[i - row][j - col] == 1) {
                laptop[i][j] = 1
            }
        }
    }
}

private fun countAttachedBlocks(laptop: Array<IntArray>): Int {
    var count = 0
    for (i in 0 until N) {
        for (j in 0 until M) {
            if (laptop[i][j] == 1) {
                count++
            }
        }
    }
    return count
}
